### ISSUE TITLE
Is the `types::LongString` really only supposed to contain a valid UTF-8 data?

### ISSUE THREAD

**Issue by RGafiyatullin** [amqp-rs/amq-protocol#32]
According to the [AMQP-0.9.1 4.2.5.3](https://www.rabbitmq.com/resources/specs/amqp0-9-1.pdf), "long strings can contain any data".

> AMQP strings are variable length and represented by an integer length followed by zero or more octets of data. AMQP defines two native string types:
> * Short strings, stored as an 8-bit unsigned integer length followed by zero or more octets of data. Short strings can carry up to 255 octets of UTF-8 data, but may not contain binary zero octets.
> * Long strings, stored as a 32-bit unsigned integer length followed by zero or more octets of data. Long strings can contain any data.

Therefore according to the standard, the following fields may contain whichever data some particular authentication method sees fit:
- in `protocol::connection::Secure`, field `challenge: LongString`;
- in `protocol::connection::SecureOk`, field `response: LongString`.

The existing API makes little sense:
* it provides an opportunity to construct an instance of `LongString` from a valid UTF-8 string only (`impl From<Stiring> for LongString` and `impl From<&str> for LongString`);
* the protocol parser (`types::parsing::parse_long_string`) uses `String::from_utf8_lossy`;
* in order to get a string value from a `LongString` one should use `as_bytes(&self) -> &[u8]`, and then do the `std::str` gymnastics, making sure the returned bytes are okay to be used as a string.

**TL;DR;**
Would it hurt much if the `LongString` would be constructed from anything that is `Into<Vec<u8>>` ?

```diff
-
-impl From<String> for LongString {
-    fn from(s: String) -> Self {
-        Self(s.into_bytes())
-    }
-}
-
-impl From<&str> for LongString {
-    fn from(s: &str) -> Self {
-        s.to_owned().into()
-      }
-}
+ 
+impl<B> From<B> for LongString where B: Into<Vec<u8>> {
+    fn from(bytes: B) -> Self {
+        Self(bytes.into())
+     }
+}
```

**Comment by Keruspe** [comment]
Iirc that's a rabbitmq restriction

**Comment by Keruspe** [comment]
Note that in the beginning this is what we did, which triggered really weird bugs in production for some users of the library which is why we added the restriction

**Comment by Keruspe** [comment]
Section 16 of https://www.rabbitmq.com/amqp-0-9-1-errata.html might be relevant

**Comment by Keruspe** [comment]
Ok, I think I got it.

We used to have both short and long string contain Vec<u8>, then we switched to String because of the mentioned issues.
After that, it was pointed out that long string can actually carry Vec<u8> so it was changed, but the conversion bits were left behind.

I will double check that's what happened and then your change can probably go in

**Comment by RGafiyatullin** [comment]
I am not a real RabbitMQ specialist, but

[The value is just constructed out of any binary](https://github.com/rabbitmq/rabbitmq-server/blob/0ab453e09dab526479e040f3de6bf2f28b5a4b7f/deps/rabbit_common/codegen.py#L429]

[The binary of specified length is just taken out of the frame data](https://github.com/rabbitmq/rabbitmq-server/blob/0ab453e09dab526479e040f3de6bf2f28b5a4b7f/deps/rabbit_common/codegen.py#L163)

The generated output of codegen.py shows no UTF-8 enforcement whatsoever for the `longstr` fields:
```erlang
decode_method_fields('connection.secure', <<F0Len:32/unsigned, F0:F0Len/binary>>) ->
  #'connection.secure'{challenge = F0};
decode_method_fields('connection.secure_ok', <<F0Len:32/unsigned, F0:F0Len/binary>>) ->
  #'connection.secure_ok'{response = F0};
```

**Comment by RGafiyatullin** [comment]
While I was doing my "research" you've commented a lot :)

**Comment by RGafiyatullin** [comment]
Indeed it seems as if there is nothing to be wary about.

BTW, why does a separate ByteArray exist?

**Comment by RGafiyatullin** [comment]
Oh, the PR would need to fix the nom-parsing too (just collect bytes, no need for utf8_lossy decoding)

**Comment by michaelklishin** [comment]
AMQP 0-9-1 and RabbitMQ have certain (not valid UTF-8 but rather a subset of ASCII) restrictions on queue, exchange, routing key, policy, runtime parameter, virtual host names.

Payloads are just opaque streams of bytes. A vector of `u8`s should be perfectly acceptable. Moreover, except for the `basic.publish`, content and body method framing, the payload is never verified by RabbitMQ itself.

**Comment by Keruspe** [comment]
Thanks for the confirmation, appreciated

**Comment by Keruspe** [comment]
This got fixed a long time ago
